### PR TITLE
Web-Viewer: Don't auto-connect to `wss://hostname` when an `?url=` is missing

### PR DESCRIPTION
**Pull request by jprochazk**
### What

* Closes https://github.com/rerun-io/rerun/issues/3276
* Closes https://github.com/rerun-io/rerun/issues/3196

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3345) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3345)
- [Docs preview](https://rerun.io/preview/ee3c7c17b884bf4940bd853ef4744600f2555b0e/docs) 
- [Examples preview](https://rerun.io/preview/ee3c7c17b884bf4940bd853ef4744600f2555b0e/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/re_viewer/src/web.rs [hunk 1]
```diff
@@ -100,6 +104,15 @@ fn create_app(cc: &eframe::CreationContext<'_>, url: Option<String>) -> crate::A
...
 
+    let mut app = crate::App::new(build_info, &app_env, startup_options, re_ui, cc.storage);
+
+    if url.is_none() && !fallback_to_ws {
+        // We didn't get a `url` to connect to, and we also aren't supposed to
+        // start the WebSocket client, so just show the welcome screen.
+        return app;
+    }
+
+    let url = url.unwrap_or_else(|| get_url(&cc.integration_info));
```
Comment (emilk) [inline crates/re_viewer/src/web.rs:115]: This looks wrong. `fallback_to_ws` will now affect wether or not we check for the `?url=…` query in `get_url`

File: web_viewer/index.html [hunk 2]
```diff
@@ -328,8 +328,12 @@
 
             check_for_panic();
 
-            let url = determine_url();
-            handle.start("the_canvas_id", url).then(on_app_started).catch(on_wasm_error);
+            let url = new URLSearchParams(window.location.search).get("url");
+            let fallback_to_websockets = false; // We're on `app.rerun.io`, so don't fallback to WebSockets.
```
Comment (emilk) [inline web_viewer/index.html:332]: I believe `web_viewer/index.html` is used in more places than just `app.rerun.io`, including just running `cargo rerun-web` and when calling `serve` from our logging SDK, e.g. when running `cargo r -p raw_mesh -- --serve`

File: crates/re_viewer/src/web.rs (hunk 1 above)
Comment (jprochazk) [reply crates/re_viewer/src/web.rs:115]: I wonder if we actually ever hit the `get_url` path, though?

File: web_viewer/index.html [hunk 3]
```diff
@@ -345,37 +349,6 @@
...
-            // index.html stripped off if it's present.
-            // exa: 'https://app.rerun.io/version/v4.0.0/index.html' -> '/data/version/v4.0.0/'
-            let data_path = '/data/' + window.location.pathname.replace(/index\.html$/, '') + '/';
-
-            const rrd_file = url_params.get('file') || 'colmap_fiat.rrd';
-
-            // Normalize the extra slashes from the url
-            const normalized = (data_path + rrd_file).replace(/\/{2,}/g, '/');
-            window.location.search = `?url=${normalized}`;
-            return normalized;
```
Comment (emilk) [inline web_viewer/index.html:376]: Removing this code is definitely the right choice

File: web_viewer/index.html (hunk 2 above)
Comment (jprochazk) [reply web_viewer/index.html:332]: When I run `cargo rerun-web` or `cargo r -p raw_mesh -- --serve`, it sends me to `http://localhost:9090/?url=ws://localhost:9877`, which works as expected.

**Review by emilk — COMMENTED** [review]
I'm not sure this is 100% the right approach.

The `fallback_to_websockets: bool` seems to me quite a hack, especially as it is controlled from JavaScript. It is, as far as I can tell, trying to subvert the fallback in `get_url` for what URL to connect to. Do we need this fallback at all though? All the places where we open the web-viewer for the user we already add the `?url=…` query parameter for them.

For instance, `cargo r -p raw_mesh -- --serve` will print this:

> Listening for websocket traffic on ws://localhost:9877. Connect with a Rerun Web Viewer.
> Started web server on http://localhost:9090
> Web server is running - view it at http://localhost:9090?url=ws://localhost:9877

and then it opens that last URL.

I think it would make more sense to remove the fallback from `get_url` and handle no URL by simply adding no receiver to the application. This will make the web-viewer without an `url=` query act similar to just running `rerun`, i.e. make it more similar to the native viewer

**Comment by jprochazk** [comment]
I removed the `fallback_to_ws` path, and changed it to only add the receiver if `url` is present.

File: crates/re_viewer/src/web.rs [hunk 4]
```diff
@@ -173,38 +185,22 @@ enum EndpointCategory {
...
+        } else {
+            EndpointCategory::WebSocket(uri.into())
         }
-        EndpointCategory::WebSocket(uri)
-    }
-}
-
-fn get_url(info: &eframe::IntegrationInfo) -> String {
-    let mut url = String::new();
-    if let Some(param) = info.web_info.location.query_map.get("url") {
```
Comment (emilk) [inline crates/re_viewer/src/web.rs:196]: So we no longer check for the `?url=…` query parameter in Rust now - what are the consequences of this?

Comment (jprochazk) [reply crates/re_viewer/src/web.rs:196]: We do, I just moved it

Comment (emilk) [reply crates/re_viewer/src/web.rs:196]: oh I see!

**Comment by emilk** [comment]
Should we also remove this dropdown in this PR then?

image

The demo link in the PR description https://demo.rerun.io/pr/3345 does not show the welcome page

**Comment by jprochazk** [comment]
I think getting rid of `demo.rerun.io` is a separate step, as is cleaning up all the related scripts and CI to not do redundant work anymore

File: crates/re_viewer/src/web.rs [hunk 5]
```diff
@@ -173,38 +185,22 @@ enum EndpointCategory {
...
     } else if uri.starts_with("web_event:") {
         EndpointCategory::WebEventListener
     } else {
         // If this is sometyhing like `foo.com` we can't know what it is until we connect to it.
         // We could/should connect and see what it is, but for now we just take a wild guess instead:
         re_log::info!("Assuming WebSocket endpoint");
         if !uri.contains("://") {
-            uri = format!("{}://{uri}", re_ws_comms::PROTOCOL);
+            EndpointCategory::WebSocket(format!("{}://{uri}", re_ws_comms::PROTOCOL))
+        } else {
```
Comment (emilk) [inline crates/re_viewer/src/web.rs:201]: nit: this if-else can we switched to avoid double-negation

**Review by emilk — APPROVED** [review]

**Pull request merged**